README.md
# top_bottom_sheet_flutter

# About
    -This plugin will help you to show the bottom sheet to the top of the screen
## Parameters
- isShowCloseButton
- closeButtonRadius
- closeButtonBackgroundColor
- closeButtonIcon
- child
- context

## Installation

Install the dependencies and devDependencies and start the server.

```sh
flutter pub add top_bottom_sheet_flutter
flutter pub get
```
Or

```sh
dependencies:
  top_bottom_sheet_flutter: $currentVersion$
```
# Example

```sh
  void showSheet() {
    TopModalSheet.show(
        context: context,
        isShowCloseButton: true,
        closeButtonRadius: 20.0,
        closeButtonBackgroundColor: Colors.white,
        child: Container(
          color: Colors.white,
          child: SizedBox(
            height: MediaQuery.of(context).size.height / 2,
            child: ListView.separated(
              itemBuilder: (BuildContext context, int index) {
                return ListTile(
                  title: const Text('Ankit Tiwari'),
                  subtitle: const Text('tiwariankit496@gmail.com'),
                  leading: FloatingActionButton(
                    heroTag: index,
                    backgroundColor: Colors.white,
                    onPressed: () {},
                    child: const FlutterLogo(),
                  ),
                );
              },
              itemCount: 30,
              separatorBuilder: (BuildContext context, int index) {
                return const Divider();
              },
            ),
          ),
        ));
  }
```

# Screenshot

![Screenshot_2022_06_11_15_38_01_686_com_example_test_demo_ankit](https://user-images.githubusercontent.com/54878509/173183909-bbfdabe0-9644-444e-aa99-029c1e06cfb9.jpg)
![Screenshot_2022_06_11_15_37_19_088_com_example_test_demo_ankit](https://user-images.githubusercontent.com/54878509/173183910-62f1977f-a4ff-4d33-9cc9-746d64f9a5fe.jpg)

# How to reach us
- tiwariankit496@gmail.com
- https://github.com/ankittiwari25



















top_bottom_sheet_flutter/CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2
 Example added

top_bottom_sheet_flutter/test/top_bottom_sheet_flutter_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:top_bottom_sheet_flutter/top_bottom_sheet_flutter.dart';

void main() {
  test('adds one to input values', () {
    const TopModalSheet(child:Text("dfdf"));
  });
}


top_bottom_sheet_flutter/lib/top_bottom_sheet_flutter.dart
library top_bottom_sheet_flutter;

import 'package:flutter/widgets.dart';

/// Top Bottom Sheet
import 'dart:async';

import 'package:flutter/material.dart';

class TopModalSheet extends StatefulWidget {
  final Color closeButtonBackgroundColor;
  final Widget? closeButtonIcon;
  final double closeButtonRadius;
  final bool isShowCloseButton;
  final Widget child;

  const TopModalSheet(
      {Key? key,
        required this.child,
        this.closeButtonBackgroundColor = Colors.white,
        this.closeButtonIcon,
        this.closeButtonRadius = 20.0,
        this.isShowCloseButton = false})
      : super(key: key);

  @override
  _TopModalSheetState createState() => _TopModalSheetState();

  static show(
      {required BuildContext context,
        @required child,
        closeButtonIcon,
        closeButtonRadius,
        isShowCloseButton,
        closeButtonBackgroundColor}) {
    Navigator.push(
        context,
        PageRouteBuilder(
            pageBuilder: (_, __, ___) {
              return TopModalSheet(
                child: child,
                closeButtonBackgroundColor:
                closeButtonBackgroundColor ?? Colors.white,
                closeButtonIcon: closeButtonIcon,
                closeButtonRadius: double.parse(closeButtonRadius.toString()),
                isShowCloseButton: isShowCloseButton,
              );
            },
            opaque: false));
  }
}

class _TopModalSheetState extends State<TopModalSheet>
    with SingleTickerProviderStateMixin {
  late Animation<double> _animation;
  late AnimationController _animationController;

  final GlobalKey _childKey = GlobalKey();

  dynamic get _childHeight {
    final renderBox = _childKey.currentContext!.findRenderObject() as RenderBox;
    return renderBox.size.height;
  }

  bool get _dismissUnderway =>
      _animationController.status == AnimationStatus.reverse;

  @override
  void initState() {
    super.initState();

    _animationController = AnimationController(
        vsync: this, duration: const Duration(milliseconds: 200));

    _animation = Tween<double>(begin: _isDirectionTop() ? -1 : 1, end: 0)
        .animate(_animationController);

    _animationController.addStatusListener((status) {
      if (status == AnimationStatus.dismissed) Navigator.pop(context);
    });

    _animationController.forward();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void _handleDragUpdate(DragUpdateDetails details) {
    if (_dismissUnderway) return;

    var change =
        details.primaryDelta! / (_childHeight ?? details.primaryDelta!);
    if (_isDirectionTop()) {
      _animationController.value += change;
    } else {
      _animationController.value -= change;
    }
  }

  void _handleDragEnd(DragEndDetails details) {
    if (_dismissUnderway) return;

    if (details.velocity.pixelsPerSecond.dy > 0 && _isDirectionTop()) return;
    if (details.velocity.pixelsPerSecond.dy < 0 && !_isDirectionTop()) return;

    if (details.velocity.pixelsPerSecond.dy > 700) {
      final double flingVelocity =
          -details.velocity.pixelsPerSecond.dy / _childHeight;
      if (_animationController.value > 0.0) {
        _animationController.fling(velocity: flingVelocity);
      }
    } else if (_animationController.value < 0.5) {
      if (_animationController.value > 0.0) {
        _animationController.fling(velocity: -1.0);
      }
    } else {
      _animationController.reverse();
    }
  }

  @override
  Widget build(BuildContext context) {
    var width = MediaQuery.of(context).size.width;
    return WillPopScope(
        onWillPop: onBackPressed,
        child: GestureDetector(
          onVerticalDragUpdate: _handleDragUpdate,
          onVerticalDragEnd: _handleDragEnd,
          child: SafeArea(
            child: Scaffold(
              backgroundColor: Colors.black38,
              body: Stack(
                children: [
                  SingleChildScrollView(
                    child: Column(
                      key: _childKey,
                      children: <Widget>[
                        AnimatedBuilder(
                            animation: _animation,
                            builder: (context, _) {
                              return Transform(
                                transform: Matrix4.translationValues(
                                    0.0, width * _animation.value, 0.0),
                                child: SizedBox(
                                  width: width,
                                  child: GestureDetector(
                                      behavior: HitTestBehavior.opaque,
                                      onTap: () {},
                                      child: SizedBox(child: widget.child)),
                                ),
                              );
                            }),
                        const SizedBox(height: kToolbarHeight),
                      ],
                    ),
                  ),
                  if (_isDirectionTop())
                    Visibility(
                      visible: widget.isShowCloseButton,
                      child: Positioned(
                        bottom: 0,
                        child: GestureDetector(
                          onTap: () => _animationController.reverse(),
                          child: Container(
                              height: kToolbarHeight,
                              width: MediaQuery.of(context).size.width,
                              color: Colors.transparent,
                              child: Center(child: closeButtonIcon())),
                        ),
                      ),
                    )
                ],
              ),
            ),
          ),
          excludeFromSemantics: true,
        ));
  }

  Widget? closeButtonIcon() {
    if (widget.closeButtonIcon == null) {
      return CircleAvatar(
        radius: double.parse(widget.closeButtonRadius.toString()),
        backgroundColor: widget.closeButtonBackgroundColor,
        child: const Icon(Icons.close),
      );
    } else {
      return widget.closeButtonIcon;
    }
  }

  bool _isDirectionTop() {
    return true;
  }

  Future<bool> onBackPressed() {
    _animationController.reverse();
    return Future<bool>.value(false);
  }
}

enum TopModalSheetDirection { top, bottom }


top_bottom_sheet_flutter/example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:top_bottom_sheet_flutter/top_bottom_sheet_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Top Bottom Sheet'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  void _showSheet() {
    TopModalSheet.show(
        context: context,
        isShowCloseButton: true,
        closeButtonRadius: 20.0,
        closeButtonBackgroundColor: Colors.white,
        child: Container(
          color: Colors.white,
          child: SizedBox(
            height: MediaQuery.of(context).size.height / 2,
            child: ListView.separated(
              itemBuilder: (BuildContext context, int index) {
                return ListTile(
                  title: const Text('Ankit Tiwari'),
                  subtitle: const Text('tiwariankit496@gmail.com'),
                  leading: FloatingActionButton(
                    heroTag: index,
                    backgroundColor: Colors.white,
                    onPressed: () {},
                    child: const FlutterLogo(),
                  ),
                );
              },
              itemCount: 30,
              separatorBuilder: (BuildContext context, int index) {
                return const Divider();
              },
            ),
          ),
        ));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
            child: ElevatedButton(
                onPressed: _showSheet, child: const Text('Show'))));
  }
}
